File: utbot-framework/src/main/java/org/utbot/quickcheck/generator/java/lang/ByteGenerator.java
/*
 The MIT License

 Copyright (c) 2010-2021 Paul R. Holser, Jr.

 Permission is hereby granted, free of charge, to any person obtaining
 a copy of this software and associated documentation files (the
 "Software"), to deal in the Software without restriction, including
 without limitation the rights to use, copy, modify, merge, publish,
 distribute, sublicense, and/or sell copies of the Software, and to
 permit persons to whom the Software is furnished to do so, subject to
 the following conditions:

 The above copyright notice and this permission notice shall be
 included in all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

package org.utbot.quickcheck.generator.java.lang;

import org.utbot.engine.greyboxfuzzer.util.UtModelGenerator;
import org.utbot.framework.plugin.api.UtModel;
import org.utbot.quickcheck.generator.GenerationStatus;
import org.utbot.quickcheck.generator.InRange;
import org.utbot.quickcheck.generator.IntegralGenerator;
import org.utbot.quickcheck.internal.Comparables;
import org.utbot.quickcheck.random.SourceOfRandomness;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Collections;
import java.util.function.Function;
import java.util.function.Predicate;

import static org.utbot.external.api.UtModelFactoryKt.classIdForType;
import static org.utbot.quickcheck.internal.Reflection.defaultValueOf;
import static java.util.Arrays.asList;

/**
 * Produces values of type {@code byte} or {@link Byte}.
 */
public class ByteGenerator extends IntegralGenerator<Byte> {
    private byte min = (Byte) defaultValueOf(InRange.class, "minByte");
    private byte max = (Byte) defaultValueOf(InRange.class, "maxByte");

    public ByteGenerator() {
        super(Collections.singletonList(Byte.class));
    }

    /**
     * Tells this generator to produce values within a specified minimum and/or
     * maximum, inclusive, with uniform distribution.
     *
     * {@link InRange#min} and {@link InRange#max} take precedence over
     * {@link InRange#minByte()} and {@link InRange#maxByte()}, if non-empty.
     *
     * @param range annotation that gives the range's constraints
     */
    public void configure(InRange range) {
        min =
            range.min().isEmpty()
                ? range.minByte()
                : Byte.parseByte(range.min());
        max =
            range.max().isEmpty()
                ? range.maxByte()
                : Byte.parseByte(range.max());
    }

    @Override public UtModel generate(
        SourceOfRandomness random,
        GenerationStatus status) {

        return UtModelGenerator.getUtModelConstructor().construct(random.nextByte(min, max), classIdForType(Byte.class));
    }

    @Override protected Function<BigInteger, Byte> narrow() {
        return BigInteger::byteValue;
    }

    @Override protected Predicate<Byte> inRange() {
        return Comparables.inRange(min, max);
    }

    @Override protected Byte leastMagnitude() {
        return Comparables.leastMagnitude(min, max, (byte) 0);
    }

    @Override protected boolean negative(Byte target) {
        return target < 0;
    }

    @Override protected Byte negate(Byte target) {
        return (byte) -target;
    }

    @Override public BigDecimal magnitude(Object value) {
        return BigDecimal.valueOf(narrow(value));
    }
}

File: utbot-framework/src/main/kotlin/org/utbot/quickcheck/generator/java/lang/ByteGenerator.kt
package org.utbot.quickcheck.generator.java.lang

import org.utbot.quickcheck.generator.GeneratorContext
import org.utbot.external.api.classIdForType
import org.utbot.framework.plugin.api.UtModel
import org.utbot.framework.plugin.api.util.byteWrapperClassId
import org.utbot.quickcheck.generator.GenerationStatus
import org.utbot.quickcheck.generator.InRange
import org.utbot.quickcheck.generator.IntegralGenerator
import org.utbot.quickcheck.internal.Reflection
import org.utbot.quickcheck.random.SourceOfRandomness

/**
 * Produces values of type `byte` or [Byte].
 */
class ByteGenerator : IntegralGenerator(listOf(Byte::class.javaObjectType)) {
    private var min = Reflection.defaultValueOf(InRange::class.java, "minByte") as Byte
    private var max = Reflection.defaultValueOf(InRange::class.java, "maxByte") as Byte

    /**
     * Tells this generator to produce values within a specified minimum and/or
     * maximum, inclusive, with uniform distribution.
     *
     * [InRange.min] and [InRange.max] take precedence over
     * [InRange.minByte] and [InRange.maxByte], if non-empty.
     *
     * @param range annotation that gives the range's constraints
     */
    fun configure(range: InRange) {
        min = if (range.min.isEmpty()) range.minByte else range.min.toByte()
        max = if (range.max.isEmpty()) range.maxByte else range.max.toByte()
    }

    override fun generate(
        random: SourceOfRandomness,
        status: GenerationStatus
    ): UtModel {
        return generatorContext.utModelConstructor.construct(random.nextByte(min, max), byteWrapperClassId)
    }
}